components/tidb_query_common/src/storage/scanner.rs

```diff
@@ -23,6 +33,34 @@ pub struct RangesScanner<T> {
     current_range: IntervalRange,
     working_range_begin_key: Vec<u8>,
     working_range_end_key: Vec<u8>,
+    rescheduler: RescheduleChecker,
+}
+
+// TODO: maybe it's better to make it generic to avoid directly depending
+// on yatp's rescheduler.
+struct RescheduleChecker {
+    prev_start: Instant,
+    prev_key_count: usize,
+}
+
+impl RescheduleChecker {
+    fn new() -> Self {
+        Self {
+            prev_start: Instant::now(),
+            prev_key_count: 0,
+        }
+    }
+
+    async fn check_reschedule(&mut self, force_check: bool) {
+        self.prev_key_count += 1;
+        if (force_check || self.prev_key_count % CHECK_KEYS == 0)
+            && self.prev_start.saturating_elapsed() > MAX_TIME_SLICE
+        {
+            reschedule().await;
```

Comment:
I'm not sure, but if we indeed want a reschedule check every 32 rows (considering that every 1024 rows are too rough), how about simply reducing the batch size?

Reply:
I have tested adjusting the batch-size based on the duration used by each poll, the performance is better in some case but the duration of one poll can still be very large(> 10ms) and the performance impact is still very large. I think one of the root cause is that if the iterator contains a lot of small ranges, it need to rebuild the scanner/iterator after each small range,  the is much more costly. 
BTW, in the future we may push down the reschedule to a lower level, or we may eventually let rocksdb support async io and eliminate this kind of manually reschedule, either way, this should be a pre-step to do so.